Overhaul Makefile for generated code

thockin opened the pull request:
Calculate all go dependencies in a single Go program, rather than a complicated series of `go list` invocations and `Makefile` logic.  This simplifies the
Makefile dramatically.

Also adds tests to ensure the dependencies stay correct.  This test is pretty slow, though.

Also fixes a few anomalies found by the test.

Fixes #54371

```release-note
NONE
```

munnerz commented:
Whilst this does simplify the source Makefile a fair bit, it does mean we're now generating a Makefile through string manipulation which is then used to generate code.

I'm not by any means saying that a massive complex Makefile is better, and the addition of tests to this area of the codebase is a great thing, but I do wonder if generating an actual Makefile is the best way to do this. Would it make more sense to maybe make the `go2make` tool a bit more generic, and perhaps just output a list of things to process or something?

I'm just thinking it'd be a lot easier for a newcomer to understand "this tool outputs a list of strings that need to be processed" than "this Makefile is half complete, and this tool generates the remainder of it when run as part of the CI process in some other script somewhere else in the repo".

You definitely know better than I do in these parts though, so please do disregard all of this message if you disagree!

/cc @sttts

thockin commented:
I considered going to full Makefile emission, but that would be the opposite of making it a more general tool.  It would mean that any changes to codegen logic would have to be written in Go to emit Makefile code, mostly as literals or templated strings.  

The goal of this is to retain the nice "do nothing if nothing needs to be done" properties, while simplifying it for people (other than me) to manage.  I am very much open to discussing the details thereof.  This is a first shot.

thockin commented:
/retest

thockin commented:
@ixdy This says a test failed, but it doesn't show me WHY it failed (and of course, it doesn't fail locally).  What is breaking down that the actual failure is not coming through?

thockin commented:
/retest

thockin commented:
@ixdy Bazel test failing to find ./testdata -- I am at wits end with Bazel.  Help?

ixdy commented on hack/make-rules/helpers/go2make/BUILD:
```diff
@@ -0,0 +1,44 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")
```
I think you want to add a line
```
# gazelle:exclude testdata
```
to this BUILD file, delete all BUILD files under testdata which have been generated, and then re-run `hack/update-bazel.sh`.

you might then need to re-add the `data = glob(["testdata/**"]),` line to the `go_test` here; I'm not sure how `gazelle:exclude` affects the data attributes, if at all.

fejta replied:
How do you see `gazelle:exclude testdata` part being helpful?

ixdy commented:
/assign @spxtr 
who says he'd also like to take a look at this

spxtr commented on hack/make-rules/helpers/go2make/go2make.go:
```diff
@@ -0,0 +1,164 @@
+	debug("load", pkgName)
+	pkg, err := build.Import(pkgName, getwd(), 0)
+	if err != nil {
+		// We can ignore NoGoError.  Anything else is real.
```
Is `pkg` nil on that error?

thockin replied:
No. NoGoError returns a valid value *and* an error.

spxtr commented on hack/verify-generated-files-remake.sh:
```diff
@@ -21,10 +21,328 @@ set -o pipefail
+
+assert_clean
+
+DIR=staging/src/k8s.io/sample-apiserver/pkg/apis/wardle/v1alpha1
```
I'm worried that this file will either get out of hand or else be neglected. You're adding over 300 lines of bash here. Is it possible to get this sort of coverage with go tests?

Related: if this test passes, do you ever expect the next to fail?

thockin replied:
I waffled on whether to test all generators, or just test one and assume the rest are the same.  I err on the side of caution.

Bash is not, of itself, a bad thing.  Fundamentally this is driving the calls of `make` commands.  Writing it in go is possible, but seems artificial.  Given that we have NOTHING covering this today, I'd rather go with this than nothing.  I don't have time to rewrite it right now :(

At this point we should probably wait until after we reopen for 1.10 development to merge this (if we do).

spxtr replied:
> I waffled on whether to test all generators, or just test one and assume the rest are the same. I err on the side of caution.

My concern is that it won't stay up-to-date, so we'll be testing most generators, not all.

thockin replied:
most is better than none?  @pwittrock and @sttts and I were discussing maybe reducing this to a single mondo-generator, which would simplify..

spxtr reviewed (commented):
Overall looks great.

thockin commented on hack/make-rules/helpers/go2make/go2make.go:
```diff
@@ -0,0 +1,164 @@
+		}
+
+		// Emit a variable for each package.
+		var buf bytes.Buffer
```
Can we followup on that?  I want to get this off my plate or kill it.

spxtr replied:
Sure, it's not too complicated.

fejta replied:
LOL Tim's comment

thockin commented:
/retest

thockin commented:
It's green now, so I know it WORKS.  I am not going to touch this again until 1.10 dev opens.

spxtr commented:
@thockin do you still plan on moving this forward?

spiffxp commented:
/unassign

thockin committed:
Get rid of verify_generated_files

It doesn't work right, we autogenerate on-demand, and we want to stop
checking them in.

thockin committed:
Build: Test if .todo file exists AND is empty

thockin committed:
Simplify Makefile for genfiles a bit

thockin committed:
Better test for generated file rebuilds

thockin committed:
Simplify find in cache_go_dirs.sh

thockin committed:
Add a helper tool to emit Makefile snippets

This loads the whole go universe one time, so should be faster that
repeated calls to `go list`

thockin committed:
Simplify build for deepcopy

thockin committed:
Simplify build for defaulter

thockin committed:
Simplify build for conversion

thockin committed:
Simplify build for openapi

thockin committed:
Simplify build for bindata

thockin commented:
I have re-tooled and rebased this.  @spiffxp 

@fejta please consider this WRT your bazel changes - this is much much simpler, IMO.  And faster.

A do-nothing rebuild is reduced from 4-ish seconds to 2-is seconds on my laptop.

dims commented:
looks like `hack/make-rules/BUILD` needs an update

thockin commented:
@fejta can you show me what sort of output you would want?

thockin commented:
Bazel updated and re-pushed.  Png @fejta

fejta commented:
/unassign @spxtr 
/assign

Sorry I tend to forget about anything that isn't on my PR dashboard -- and I try to get to everything on my dashboard at least once a day. Will look soon

fejta commented:
Will also wait for the PR to pass tests.

thockin commented:
@fejta guidance on bazel failure?  It's not finding testdata dir, it seems.

thockin commented:
> Failed to delete output files after incomplete download. Cannot continue with local execution.

This is why we vendor EVERYTHING - is bazel trying to download stuff?

/retest

And why is bazel testing not automatically including testdata/ directories?  That seems wrong...

thockin commented:
Does anyone else get hit with a new bazel bug on every PR, or is it just me? :)

ixdy commented:
well, that one intermittently hits everyone, though you do seem to attract weird bugs. :)

/test pull-kubernetes-e2e-gce-device-plugin-gpu

fejta commented on hack/make-rules/helpers/go2make/BUILD:
```diff
@@ -0,0 +1,41 @@
+
+go_test(
+    name = "go_default_test",
+    srcs = ["pkgwalk_test.go"],
```
Your test is failing because it is trying to open a ./testdata/dir-without-gofiles which doesn't exist.

Can you try adding `data = [":package-srcs"]` here? You can also try the `data = glob(["testdata/**"])`, but that's included in package-srcs.

Although I suspect it might make the dir hack/make-rules/helpers/go2make/testdata

thockin commented:
/retest

thockin commented:
@fejta @ixdy all green

ixdy commented on build/root/Makefile.generated_files:
```diff
@@ -37,81 +37,54 @@ SHELL := /bin/bash
+# the package - for example the variable for `k8s.io/kubernetes/pkg/api` is
+# $(k8s.io/kubernetes/pkg/api).  This is roughly the same DAG that the Go
+# compiler uses.  These variables can be used to figure out if, for example,
+# generated code needs to be regerated.
```
nit, minor typo: regerated -> regenerated

ixdy commented on build/root/Makefile.generated_files:
```diff
@@ -37,81 +37,54 @@ SHELL := /bin/bash
+# Include the Go package dependencies file.  This will cause the rule of
+# the same name to be considered and if it is updated, make will restart and
+# reload the updated deps.
+sinclude $(META_DIR)/$(GO_PKGDEPS_FILE)
```
is this including the `$(GO_PKGDEPS_FILE)` before it's updated? or is `make` smart enough to generate this file first, since there's a rule for it?

thockin replied:
Make will try to regenerate it before including it.

ixdy commented on build/root/Makefile.generated_files:
```diff
@@ -736,22 +425,16 @@ gen_openapi: $(OPENAPI_OUTFILE) $(OPENAPI_GEN)
 # How to regenerate open-api code.  This emits a single file for all results.
-$(OPENAPI_OUTFILE): $(OPENAPI_GEN) $(OPENAPI_GEN)
+$(OPENAPI_OUTFILE): $(OPENAPI_GEN)
 	function run_gen_openapi() {                                                        \
```
is there a reason to keep this shell function here (which gets immediately called)? it seems like you unrolled similar functions in the other cases.

thockin replied:
nope, fixed.

ixdy reviewed (commented):
Generally LG, just a few questions/comments.

I haven't looked at `hack/verify-generated-files-remake.sh` yet but wanted to post my pending comments.

ixdy commented:
> Ideally what I would like is a single go program which lets me send it a list of packages and then it tells me the go list -f {{.Deps}} for each package, which I can then | sort -u to get an overall dep list.
>
> Any chance we can put that option into here?

@fejta I'm not sure how useful `go2make` is for the bazel workflow. It's basically a simplified version of gazelle, except that it prints out make rules instead of build rules. (We also already have the list of deps in each package encoded into the `go_library` deps for that package.)

thockin commented on hack/make-rules/helpers/cache_go_dirs.sh:
```diff
@@ -39,7 +39,7 @@ function kfind() {
     # of the Kubernetes source tree - generators will use these for
     # including certain core API concepts.
-    find -H . ./vendor/k8s.io/apimachinery ./vendor/k8s.io/apiserver ./vendor/k8s.io/kube-aggregator ./vendor/k8s.io/apiextensions-apiserver ./vendor/k8s.io/metrics ./vendor/k8s.io/sample-apiserver ./vendor/k8s.io/api ./vendor/k8s.io/client-go ./vendor/k8s.io/code-generator ./vendor/k8s.io/sample-controller \
+    find -H . ./vendor/k8s.io/*    \
```
It's not really what I want, but it's not a practical problem because those don't have the files we care about.  Let me see if I can do this better

thockin replied:
fixed in next push

thockin commented on hack/make-rules/helpers/go2make/pkgwalk.go:
```diff
@@ -0,0 +1,117 @@
+var ErrSkipPkg = fmt.Errorf("package skipped")
+
+// WalkPkg recursively visits all packages under pkgName.  This is similar
+// to filepath.Walk, except that it follows symlinks.  A package is always
```
yes

thockin commented on build/root/Makefile.generated_files:
```diff
@@ -35,78 +35,58 @@ SHELL := /bin/bash
+	    echo "DBG: calculating Go dependencies";  \
+	fi
+	hack/run-in-gopath.sh go install ./hack/make-rules/helpers/go2make
+	hack/run-in-gopath.sh go2make                    \
```
```
thockin@freakshow2 kubernetes make-genfiles-cleanup /$ make clean
+++ [0531 12:45:52] Verifying Prerequisites....
+++ [0531 12:45:53] Removing _output directory
Removing _tmp ..
Removing pkg/generated/bindata.go ..
Removing test/e2e/generated/bindata.go ..
thockin@freakshow2 kubernetes make-genfiles-cleanup /$ time hack/run-in-gopath.sh go install ./hack/make-rules/helpers/go2make

real	0m0.708s
user	0m1.028s
sys	0m0.204s

thockin@freakshow2 kubernetes make-genfiles-cleanup /$ time hack/run-in-gopath.sh go install ./hack/make-rules/helpers/go2make

real	0m0.216s
user	0m0.168s
sys	0m0.116s
```

So it builds itself from scratch in 0.7 sec (on my desktop) and a do-nothing costs 0.2 sec.

Running it:

```
thockin@freakshow2 kubernetes make-genfiles-cleanup /$ time hack/run-in-gopath.sh go2make k8s.io/kubernetes/... --prune  k8s.io/kubernetes/staging --prune  k8s.io/kubernetes/vendor k8s.io/kubernetes/vendor/k8s.io/... > /dev/null

real	0m4.920s
user	0m2.556s
sys	0m1.420s

thockin@freakshow2 kubernetes make-genfiles-cleanup /$ time hack/run-in-gopath.sh go2make k8s.io/kubernetes/... --prune  k8s.io/kubernetes/staging --prune  k8s.io/kubernetes/vendor k8s.io/kubernetes/vendor/k8s.io/... > /dev/null

real	0m2.435s
user	0m2.000s
sys	0m0.860s
```

thockin commented:
all issues fixed

ixdy commented:
I think your `# gazelle:exclude testdata` may have gotten lost in rebase. :\

thockin commented:
OK, green once again.  And now I missed the boat.  Sigh.

ixdy commented:
/lgtm

sttts commented:
🎉
